src/coreclr/vm/interpexec.cpp

```diff
@@ -6,6 +6,29 @@
 #include "threads.h"
 #include "gcenv.h"
 #include "interpexec.h"
+#include "callstubgenerator.h"
+
+void InvokeCompiledMethod(MethodDesc *pMD, int8_t *pArgs, int8_t *pRet)
+{
+    CallStubGenerator callStubGenerator;
+    CallStubHeader *pHeader = pMD->GetCallStubHeader();
+    if (pHeader == NULL)
+    {
+        pHeader = callStubGenerator.GenerateCallStub(pMD);
+        HRESULT hr = pMD->SetCallStubHeader(pHeader);
+        if (hr == S_FALSE)
+        {
+            // We have lost the race for generating the header, so we need to free the one we generated
+            // and use the one that was generated by another thread.
+            CallStubGenerator::FreeCallStub(pHeader);
+            pHeader = pMD->GetCallStubHeader();
+        }
+    }
+
+    pHeader->SetTarget(pMD->GetNativeCode()); // The method to call
```

Comment:
This doesn't look thread safe, I assume that's a TODO?

Reply:
@kg, it should not be a problem. The write should be atomic and we don't really care about the version of code that will get executed. If JIT generates a new version of code, it is ok for both threads to call the newer or older one, because depending on the timing, they would do the same even if that was a JIT to JIT call.